core/src/main/java/io/smallrye/openapi/runtime/scanner/spi/AnnotationScanner.java

```diff
@@ -470,8 +471,22 @@ default void createResponseFromRestMethod(final AnnotationScannerContext context
                     produces = getDefaultProduces(context, method);
                 }
 
-                if (schema != null && schema.getNullable() == null && TypeUtil.isOptional(returnType)) {
-                    schema.setNullable(Boolean.TRUE);
+                if (schema != null && SchemaImpl.getNullable(schema) == null && TypeUtil.isOptional(returnType)) {
+                    if (schema.getType() != null) {
+                        schema.addType(Schema.SchemaType.NULL);
+                    }
+                    if (schema.getRef() != null) {
+                        Schema nullSchema = new SchemaImpl().type(singletonList(Schema.SchemaType.NULL));
+                        // Move reference to type into its own subschema
+                        Schema refSchema = new SchemaImpl().ref(schema.getRef());
+                        schema.setRef(null);
+                        if (schema.getAnyOf() == null) {
+                            schema.addAnyOf(refSchema).addAnyOf(nullSchema);
+                        } else {
+                            Schema anyOfSchema = new SchemaImpl().addAnyOf(refSchema).addAnyOf(nullSchema);
+                            schema.addAllOf(anyOfSchema);
+                        }
+                    }
                 }
```

Comment:
@MikeEdgar you were correct. The old code here did not consider that `schema` might have `$ref` set. If `$ref` was set, then value for `nullable` set here got discarded when the schema was serialized.

It looked like the intent of the old code was to add `nullable = true` when a resource method returned `Optional<T>` so I made the new code handle this case whether or not the response schema is a reference.